typescript/sdk/src/providers/transactionFeeEstimators.ts

```diff
@@ -0,0 +1,260 @@
+    from: sender,
+  });
+  const feeData = await provider.provider.getFeeData();
+  let gasPrice: BigNumber;
+  if (feeData.maxFeePerGas && feeData.maxPriorityFeePerGas) {
+    gasPrice = feeData.maxFeePerGas.add(feeData.maxPriorityFeePerGas);
+  } else if (feeData.gasPrice) {
+    gasPrice = feeData.gasPrice;
+  } else {
+    throw new Error(`Invalid fee data: ${JSON.stringify(feeData)}`);
+  }
+  return {
+    gasUnits: BigInt(gasUnits.toString()),
+    gasPrice: BigInt(gasPrice.toString()),
+    fee: BigInt(gasUnits.mul(gasPrice).toString()),
+  };
+}
+
+export async function estimateTransactionFeeViem({
+  transaction,
+  provider,
+  sender,
+}: {
+  transaction: ViemTransaction;
+  provider: ViemProvider;
+  sender: Address;
+}): Promise<TransactionFeeEstimate> {
+  const gasUnits = await provider.provider.estimateGas({
+    ...transaction.transaction,
+    blockNumber: undefined,
+    account: sender as `0x${string}`,
+  });
+  const feeData = await provider.provider.estimateFeesPerGas();
+  let gasPrice: bigint;
+  if (feeData.maxFeePerGas && feeData.maxPriorityFeePerGas) {
+    gasPrice = feeData.maxFeePerGas + feeData.maxPriorityFeePerGas;
+  } else if (feeData.gasPrice) {
+    gasPrice = feeData.gasPrice;
+  } else {
+    throw new Error(`Invalid fee data: ${JSON.stringify(feeData)}`);
+  }
+  return {
+    gasUnits,
+    gasPrice,
+    fee: gasUnits * gasPrice,
+  };
+}
+
+export async function estimateTransactionFeeSolanaWeb3({
+  provider,
+  transaction,
+}: {
+  transaction: SolanaWeb3Transaction;
+  provider: SolanaWeb3Provider;
+}): Promise<TransactionFeeEstimate> {
+  const connection = provider.provider;
+  const { value } = await connection.simulateTransaction(
+    transaction.transaction,
+  );
+  assert(!value.err, `Solana gas estimation failed: ${value.err}`);
+  const gasUnits = BigInt(value.unitsConsumed || 0);
+  const recentFees = await connection.getRecentPrioritizationFees();
+  const gasPrice = BigInt(recentFees[0].prioritizationFee);
+  return {
+    gasUnits,
+    gasPrice,
+    fee: gasUnits * gasPrice,
+  };
+}
+
+// This is based on a reverse-engineered version of the
+// SigningStargateClient's simulate function. It cannot be
+// used here because it requires access to the private key.
+// https://github.com/cosmos/cosmjs/issues/1568
+export async function estimateTransactionFeeCosmJs({
+  transaction,
+  provider,
+  estimatedGasPrice,
+  sender,
+  senderPubKey,
+  memo,
+}: {
+  transaction: CosmJsTransaction;
+  provider: CosmJsProvider;
+  estimatedGasPrice: Numberish;
+  sender: Address;
+  senderPubKey: HexString;
+  memo?: string;
+}): Promise<TransactionFeeEstimate> {
+  const stargateClient = await provider.provider;
+  const message = transaction.transaction;
+  const registry = new Registry([...defaultRegistryTypes, ...wasmTypes]);
+  const encodedMsg = registry.encodeAsAny(message);
+  const encodedPubkey = encodeSecp256k1Pubkey(Buffer.from(senderPubKey, 'hex'));
```

Comment:
can the public key be queried using the account lookup RPC? would prefer not to pollute all the interfaces with this senderPubKey

Reply:
```
What you can do is ask the chain if it stores the public key for a given address. If the address has signed transactions before, this gives you a public key on Cosmos.
```
https://github.com/cosmos/cosmjs/issues/889

Reply:
Leaving a comment about this but not fixing for now